### PR TITLE
feat(Styleguide): Convert All Stories to MDX

### PR DESCRIPTION
**Pull request by codecaaron**
# Storybook MDX
This is the full CSF => MDX switch.

Check it out: [here](https://5ebb0dd1d1cbb83db6aec644--codecademy-styleguide.netlify.app/storybook/)

Here is the related [RFC](https://www.notion.so/codecademy/Docs-first-Storybook-using-MDX-Gamut-1e44caba2c274a199db6ed895f77ecae)

## Overview
There are are a few main themes in this PR:
- **MDX Conversion**: Converts all CSF to MDX with the exception of (Tabs which is currently incompatible with MDX).
- **Updated Stories**: Added new discrete stories to many components to surface possible options.
- **Config Cleanup**: Cleans up storybook config adding baseline decorators and removing unused options.
- **File Structure**: Introduces Storybook Category / Folder Structure.
- **Storybook Blocks**: Converts existing Storybook Native components to Blocks.

## Notes
There are a few things that are currently not included in this PR:
- Full enforcement of naming standards.
- Deprecation of CSF helpers.
- Full TS / ESLint MDX support (though much of this is enabled through recommended VSCode extensions).
- There are quirks with the MDX / Storyshot integration as any generic JSX tags inside of stories will be treated like MDX components (`p`, `span`, ...etc) causing StoryShot tests to break due to "illegal hook usage".  I have an open thread on the Storybook discord to figure out what is going on, in the meantime I have converted them to named components to avoid breakages.

### MDX
The bulk of this diff is converting existing CSF story pages to MDX.  Much of this included moving the description and usage guides out of the preview window and into parsable markdown.

CSF:
```tsx
export default {
  title: 'Core|Molecules/Alert',
  component: Alert,
  decorators: [withKnobs],
};

export const alert = decoratedStory(() => (
  <LayoutGrid rowGap="xl">
    <Column size={12}>
      <StoryTemplate status={StoryStatus.Ready}>
        <StoryDescription>
          Alerts can be used to asynchronously display feedback or interactions
          with the user with various connotations.
        </StoryDescription>
      </StoryTemplate>
    </Column>

    <Column size={12}>
      <h3>Feedback Alert</h3>

      <p>
        Serves as <strong>feedback</strong> and confirmation that requires an
        action.
      </p>
      <h6>Informational/general feedback</h6>
      <Alert
        showIcon
        type={BannerType.Info}
        cta={{ text: 'Login', onClick: () => {} }}
        onClose={() => {}}
      >
        You have been logged out due to inactivity.
      </Alert>
    </Column>
</LayoutGrid>);
```
MDX:
```
<Meta title="Core|Molecules/Alert" component={Alert} />

# Alert

<StoryStatus status="ready" />

Alerts can be used to asynchronously display feedback or interactions
with the user with various connotations.

<Preview withSource="open">
  <Story name="Alert">
    <Alert
      showIcon
      type={BannerType.Info}
      cta={{ text: 'Login', onClick: action('CTA Click') }}
      onClose={action('Close')}
    >
      You have been logged out due to inactivity.
    </Alert>
  </Story>
</Preview>

<Props story="Alert" of={Alert} />
```

Base Story Order :
1. Meta Block
2. Title (h1)
2. Status Block
3. Overal Description
4. Main Story with Knobs (Preview and Story Blocks)
5. PropTable
6. Any sub stories.

Substories:
1. Title (h2)
2. Description.
3. Story Blocks

This format is not always consistent depending on the component, but for insular components it works well.  I do not as of yet have a playbook for what will work best in the future. 

### Story Updates
Some stories may have changed in various ways:
- There maybe significantly more stories now.  I have started splitting stories by discrete visual states rather than preview window.  (This is especially true in the case of[ Button](https://5ebb0dd1d1cbb83db6aec644--codecademy-styleguide.netlify.app/storybook/?path=/docs/core-atoms-button--button)).
- Added Prop Tables to components if they did not have one and added subcomponent prop tables for all commonly used together components: 
Screen Shot 2020-05-12 at 5 19 04 PM
- Converted all Icon set docs to the built in storybook icon gallery: 
Screen Shot 2020-05-12 at 5 19 43 PM

### Config
Several updates to storybook config are included in this including:
- Category / Atomic story sorting (See below section for allowed sections).
- Knobs added to all stories without needing to include it on every story.
- Standalone source formatting with `prettier` ([this is necessary for story source to be formatted correctly between MDX => CSF transpilation](https://github.com/storybookjs/storybook/issues/8078#issuecomment-605430645))
- Adds a darkmode preview classname for temporary theming support.  (This will need to be improved upon later).

### File Structure
All story files are now organized by the category they fall under inside of the `/stories/` directory.  These top level categories represent the current usage status: `Core`, `Labs`, `Deprecated`.

Inside of these directories all stories go under respective atomic design folders of: `Foundations`, `Atoms`, `Molecules`, `Organisms`.

Story titling should match directory structure, until we make them dynamic:
`${category}|${atomicClassification}/${componentName}`

### Storybook Blocks
There are a few elements that have been converted to use StoryBooks theme aware `emotion` instance including StoryStatus and Swatches. These components have been moved out of the `/stories` directory and live in their own `/templating` area.  As we develop more sophisticated storybook usages we may want to make more native components that are not tied to our design system.

### Developer Changes
1. Deploy locally and `yarn clean-modules && yarn && yarn build-all && yarn start`
2. Install recommended `vscode` extensions for `mdx` support.  I highly suggest adding prettier to your local vscode instance as well and enabling mdx formatting.
3. View any `*.stories.mdx` file and update it.

### Documentation Changes
1. Visit the [Styleguide](https://5ebb0dd1d1cbb83db6aec644--codecademy-styleguide.netlify.app/storybook/)
2. View any story and check for any breakages.

### CODE REVIEW

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5eb441bd78f002c4ea6baf14--codecademy-styleguide.netlify.app>https://5eb441bd78f002c4ea6baf14--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5eb441bd78f002c4ea6baf14>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5eb4910a322d97c5c9c5875f--codecademy-styleguide.netlify.app>https://5eb4910a322d97c5c9c5875f--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5eb4910a322d97c5c9c5875f>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5eb4c4c3c356e4f37eba001d--codecademy-styleguide.netlify.app>https://5eb4c4c3c356e4f37eba001d--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5eb4c4c3c356e4f37eba001d>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5eb4f50d3eca3b887fdd37d9--codecademy-styleguide.netlify.app>https://5eb4f50d3eca3b887fdd37d9--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5eb4f50d3eca3b887fdd37d9>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5eb5bac47276d27a2a69ebf4--codecademy-styleguide.netlify.app>https://5eb5bac47276d27a2a69ebf4--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5eb5bac47276d27a2a69ebf4>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5eb5e8d76b524c8c4b5920bb--codecademy-styleguide.netlify.app>https://5eb5e8d76b524c8c4b5920bb--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5eb5e8d76b524c8c4b5920bb>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5eb600f855780e758a01a24c--codecademy-styleguide.netlify.app>https://5eb600f855780e758a01a24c--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5eb600f855780e758a01a24c>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5eb73acf588cea0256bfb600--codecademy-styleguide.netlify.app>https://5eb73acf588cea0256bfb600--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5eb73acf588cea0256bfb600>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5eb958681842f596cf0a5b56--codecademy-styleguide.netlify.app>https://5eb958681842f596cf0a5b56--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5eb958681842f596cf0a5b56>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5eb977a21842f59fc90a5e98--codecademy-styleguide.netlify.app>https://5eb977a21842f59fc90a5e98--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5eb977a21842f59fc90a5e98>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5eb97d6edfd2a0c514e25555--codecademy-styleguide.netlify.app>https://5eb97d6edfd2a0c514e25555--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5eb97d6edfd2a0c514e25555>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5eb99c9fe52ccc54cb90c14b--codecademy-styleguide.netlify.app>https://5eb99c9fe52ccc54cb90c14b--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5eb99c9fe52ccc54cb90c14b>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5eb9a017f5037a017acad3e5--codecademy-styleguide.netlify.app>https://5eb9a017f5037a017acad3e5--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5eb9a017f5037a017acad3e5>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5eb9c4cc975cef052b13e010--codecademy-styleguide.netlify.app>https://5eb9c4cc975cef052b13e010--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5eb9c4cc975cef052b13e010>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5eb9cc5cc6dc4a807f5b0da1--codecademy-styleguide.netlify.app>https://5eb9cc5cc6dc4a807f5b0da1--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5eb9cc5cc6dc4a807f5b0da1>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5eb9d0726f64a8f2a8a0b810--codecademy-styleguide.netlify.app>https://5eb9d0726f64a8f2a8a0b810--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5eb9d0726f64a8f2a8a0b810>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5eb9e975f5037a20e8cad2bc--codecademy-styleguide.netlify.app>https://5eb9e975f5037a20e8cad2bc--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5eb9e975f5037a20e8cad2bc>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5ebadd4dbcb219ef36d37830--codecademy-styleguide.netlify.app>https://5ebadd4dbcb219ef36d37830--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5ebadd4dbcb219ef36d37830>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5ebae8575fe04601f4f5a748--codecademy-styleguide.netlify.app>https://5ebae8575fe04601f4f5a748--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5ebae8575fe04601f4f5a748>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5ebaf2c6d206f7ec14643f72--codecademy-styleguide.netlify.app>https://5ebaf2c6d206f7ec14643f72--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5ebaf2c6d206f7ec14643f72>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5ebafe38a5bd2b6811bb0a1a--codecademy-styleguide.netlify.app>https://5ebafe38a5bd2b6811bb0a1a--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5ebafe38a5bd2b6811bb0a1a>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/brand-components@1.5.14-alpha.cda79e.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5ebb0dd1d1cbb83db6aec644--codecademy-styleguide.netlify.app>https://5ebb0dd1d1cbb83db6aec644--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5ebb0dd1d1cbb83db6aec644>Deploy Logs</a></p>

File: packages/styleguide/.storybook/preview-head.html [hunk 1]
```diff
@@ -1 +1,7 @@
 <meta name="robots" content="noindex" />
+<style>
+  div.dark {
+    background-color: #19191a;
+    color: #f6f5fa;
+  }
```
Comment (codecaaron) [inline packages/styleguide/.storybook/preview-head.html:6]: necessary as this stylesheet is higher than storybooks own styles.

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5ebb17e9b9092401be3d2d3c--codecademy-styleguide.netlify.app>https://5ebb17e9b9092401be3d2d3c--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5ebb17e9b9092401be3d2d3c>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5ebb1cc5938b7b0179376971--codecademy-styleguide.netlify.app>https://5ebb1cc5938b7b0179376971--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5ebb1cc5938b7b0179376971>Deploy Logs</a></p>

File: packages/styleguide/Templating/Blocks/StoryStatus.tsx [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+const StatusIndicator = styled.div<{ color: string }>`
+  display: inline-block;
+  position: relative;
+  font-size: 1.2rem;
+  font-weight: 400;
+  width: 1rem;
+  height: 1rem;
+  border-radius: 50%;
+  margin-top: 0.2rem;
+  background-color: ${({ color }) => color};
```
Comment (JoshuaKGoldberg) [inline packages/styleguide/Templating/Blocks/StoryStatus.tsx:29]: Is there no built-in for this? Seems a little squiggle-heavy...

File: packages/styleguide/Templating/Blocks/StoryStatus.tsx [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+  &:hover {
```
Comment (JoshuaKGoldberg) [inline packages/styleguide/Templating/Blocks/StoryStatus.tsx:31]:
```suggestion
  &:focus
  &:hover, {
```

File: packages/styleguide/stories/Brand/Atoms/BrandMonospace/BrandMonospace.stories.mdx [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+import styles from './styles.module.scss';
+
+<Meta title="Labs|Atoms/BrandMonospace" component={BrandMonospace} />
+
+# BrandMonospace
+
+<StoryStatus status="inProgress" />
+
+A wrapper component that will style text with the `Robot Mono`
+font. Currently, Codecademy uses `Monaco`as our default monospaced
```
Comment (JoshuaKGoldberg) [inline packages/styleguide/stories/Brand/Atoms/BrandMonospace/BrandMonospace.stories.mdx:17]:
```suggestion
font. Currently, Codecademy uses `Monaco` as our default monospaced
```

File: packages/styleguide/stories/Brand/Atoms/BrandMonospace/BrandMonospace.stories.mdx [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+font. However, there are a few brand related instances where we are experimenting
+with using `Roboto Mono`.
+
+<Preview>
+  <Story name="BrandMonospace">
+    <>
```
Comment (JoshuaKGoldberg) [inline packages/styleguide/stories/Brand/Atoms/BrandMonospace/BrandMonospace.stories.mdx:23]: So do stories only accept a single child? Or is this unintentional?

File: packages/styleguide/stories/Brand/Atoms/Logo.stories.mdx [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+import { selectableColors } from '../../helpers';
+
+<Meta title="Labs|Atoms/Logo" component={Logo} />
+
+# Logo
+
+<StoryStatus status="ready" />
+
+Our proud logo. There is no a in the middle. Use sparingly, and make sure
+it has a lot of visual disinction to standout appropriately.
```
Comment (JoshuaKGoldberg) [inline packages/styleguide/stories/Brand/Atoms/Logo.stories.mdx:15]: The "standout" typo is new; the "disinction" one is pre-existing from me 😅 

```suggestion
it has a lot of visual distinction to stand out appropriately.
```

File: packages/styleguide/stories/Brand/Atoms/BrandMonospace/BrandMonospace.stories.mdx (hunk 5 above)
Comment (codecaaron) [reply packages/styleguide/stories/Brand/Atoms/BrandMonospace/BrandMonospace.stories.mdx:23]: Yeah, let me make of note of it to put it in the brownbag.

Screen Shot 2020-05-13 at 9 48 39 AM

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5ebbfd87f2f1f93b374dc79e--codecademy-styleguide.netlify.app>https://5ebbfd87f2f1f93b374dc79e--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5ebbfd87f2f1f93b374dc79e>Deploy Logs</a></p>

File: packages/styleguide/Templating/Blocks/StoryStatus.tsx (hunk 2 above)
Comment (codecaaron) [reply packages/styleguide/Templating/Blocks/StoryStatus.tsx:29]: I haven't dug too deep yet into this, but after a quick sanity check?  I don't think so?

In personal projects I did something along the lines of:
```tsx
import { get } from 'lodash/fp';
import styled from 'css-in-js';


const myDiv = styled.div<{ myProp: blah }>`
  blah: ${get('myProp')}
`;
```

Which cut down some things, but currying seemed like a decent strategy for more of the weird templatey things hee.

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://5ebc0106873ac06a60d060c9--codecademy-styleguide.netlify.app>https://5ebc0106873ac06a60d060c9--codecademy-styleguide.netlify.app</a></p><p><a href=https://app.netlify.com/sites/codecademy-styleguide/deploys/5ebc0106873ac06a60d060c9>Deploy Logs</a></p>

**Pull request merged**